cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+    unmatchable: Vec<UnmatchableError>,
+    rules: RuleSet,
+}
+
+impl RuleSetBuilder {
+    fn add_rule(
+        &mut self,
+        rule: &sema::Rule,
+        termenv: &sema::TermEnv,
+        tyenv: &sema::TypeEnv,
+        errors: &mut Vec<Error>,
+    ) {
+        self.current_rule.pos = rule.pos;
+        self.current_rule.prio = rule.prio;
+        self.current_rule.result = rule.visit(self, termenv);
+        self.normalize_equivalence_classes();
+        let rule = std::mem::take(&mut self.current_rule);
+
+        if self.unmatchable.is_empty() {
+            self.rules.rules.push(rule);
+        } else {
+            // If this rule can never match, drop it so it doesn't affect overlap checking.
+            errors.extend(self.unmatchable.drain(..).map(|err| {
+                let src = Source::new(
+                    tyenv.filenames[err.pos.file].clone(),
+                    tyenv.file_texts[err.pos.file].clone(),
+                );
+                Error::UnmatchableError {
+                    msg: format!(
+                        "rule requires binding to match both {:?} and {:?}",
+                        err.constraint_a, err.constraint_b
+                    ),
+                    src,
+                    span: Span::new_single(err.pos),
+                }
+            }))
+        }
+    }
+
+    // If a binding site is constrained and also required to be equal to another binding site, then
+    // copy the constraint and push the equality inside it.
+    fn normalize_equivalence_classes(&mut self) {
+        let mut constraints = Vec::new();
+        for binding in self.current_rule.equals.keys() {
+            if let Some(constraint) = self.current_rule.constraints.get(binding) {
+                constraints.push((*binding, *constraint));
+            }
+        }
+
+        while let Some((mut binding, constraint)) = constraints.pop() {
+            struct Pending {
+                binding: BindingId,
+                variant: sema::VariantId,
+                field: TupleIndex,
+            }
+            let mut bindings = if let Constraint::Variant(fields, _, variant) = constraint {
+                (0..fields)
+                    .map(|field| Pending {
+                        binding: self.dedup_binding(Binding::Variant {
+                            source: binding,
+                            variant,
+                            field,
+                        }),
+                        variant,
+                        field,
+                    })
+                    .collect()
+            } else {
+                Vec::new()
+            };
+
+            // Processing the first constraint in an equivalence class removes the equivalence
+            // class. If there were more constraints in that class, then we discard them here. But
+            // we left the original constraints in place above, so we've already checked that all
+            // the constraints are equal.
```

Comment:
This one is a bit mystifying too, sorry -- leaves a few questions for me: why remove an equivalence class after processing just one link? What do you mean by "all the constraints are equal"? Where did that check happen?